usersession/userd/settings.go

```diff
@@ -77,15 +94,116 @@ func allowedSetting(setting string) bool {
 	return snap.ValidAppName(base)
 }
 
+func settingWithSub(setting string, subproperty string) string {
+	return fmt.Sprintf("%s/%s", setting, subproperty)
+}
+
+func settingForLog(setting string) string {
+	settingParts := strings.SplitN(setting, "/", 2)
+	if len(settingParts) > 1 {
+		return fmt.Sprintf("%q subproperty %q", settingParts[0], settingParts[1])
+	} else {
+		return fmt.Sprintf("%q", settingParts[0])
+	}
+}
+
 func settingWhitelisted(setting string) *dbus.Error {
+	// Trim off any subproperty from the check
+	settingTrimmed := strings.SplitN(setting, "/", 2)[0]
 	for _, whitelisted := range settingsWhitelist {
-		if setting == whitelisted {
+		if settingTrimmed == whitelisted {
 			return nil
 		}
 	}
 	return dbus.MakeFailedError(fmt.Errorf("cannot use setting %q: not allowed", setting))
 }
 
+func safeSnapFromSender(s *Settings, sender dbus.Sender) (string, *dbus.Error) {
+	// avoid information leak: see https://github.com/snapcore/snapd/pull/4073#discussion_r146682758
+	snap, err := snapFromSender(s.conn, sender)
+	if err != nil {
+		return "", dbus.MakeFailedError(err)
+	}
+	return snap, nil
+}
+
+func desktopFile(s *Settings, command string, setting string, desktopValue string, sender dbus.Sender) (string, *dbus.Error) {
```

Comment:
this helper maybe should be called desktopFileFromValue ?   and maybe s/desktopValue/dotDesktopValue/